doc/api/quic.md

```diff
@@ -0,0 +1,2276 @@
+socket.listen({ key, cert, alpn: 'hello' });
+
+socket.on('session', (session) => {
+  // A new server side session has been created!
+
+  session.on('secure', () => {
+    // Once the TLS handshake is completed, we can
+    // open streams...
+    const uni = session.openStream({ halfOpen: true });
+    uni.write('hi ');
+    uni.end('from the server!');
+  });
+
+  // The peer opened a new stream!
+  session.on('stream', (stream) => {
+    // Let's say hello
+    stream.end('Hello World');
+
+    // Let's see what the peer has to say...
+    stream.setEncoding('utf8');
+    stream.on('data', console.log);
+    stream.on('end', () => console.log('stream ended'));
+  });
+});
+
+socket.on('listening', () => {
+  // The socket is listening for sessions!
+});
+```
+
+## QUIC Basics
+
+QUIC is a UDP-based network transport protocol that includes built-in security
+via TLS 1.3, flow control, error correction, connection migration,
+multiplexing, and more.
+
+Within the Node.js implementation of the QUIC protocol, there are three main
+components: the `QuicSocket`, the `QuicSession` and the `QuicStream`.
+
+### QuicSocket
+
+A `QuicSocket` encapsulates a binding to one or more local UDP ports. It is
+used to send data to, and receive data from, remote endpoints. Once created,
+a `QuicSocket` is associated with a local network address and IP port and can
+act as both a QUIC client and server simultaneously. User code at the
+JavaScript level interacts with the `QuicSocket` object to:
+
+* Query or modified the properties of the local UDP binding;
+* Create client `QuicSession` instances;
+* Wait for server `QuicSession` instances; or
+* Query activity statistics
+
+Unlike the `net.Socket` and `tls.TLSSocket`, a `QuicSocket` instance cannot be
+directly used by user code at the JavaScript level to send or receive data over
+the network.
+
+### Client and Server QuicSessions
+
+A `QuicSession` represents a logical connection between two QUIC endpoints (a
+client and a server). In the JavaScript API, each is represented by the
+`QuicClientSession` and `QuicServerSession` specializations.
+
+At any given time, a `QuicSession` exists is one of four possible states:
+
+* `Initial` - Entered as soon as the `QuicSession` is created.
+* `Handshake` - Entered as soon as the TLS 1.3 handshake between the client and
+  server begins. The handshake is always initiated by the client.
+* `Ready` - Entered as soon as the TLS 1.3 handshake completes. Once the
+  `QuicSession` enters the `Ready` state, it may be used to exchange
+  application data using `QuicStream` instances.
+* `Closed` - Entere as soon as the `QuicSession` connection has been
+  terminated.
+
+New instances of `QuicClientSession` are created using the `connect()`
+function on a `QuicSocket` as in the example below:
+
+```js
+const { createQuicSocket } = require('net');
+
+// Create a QuicSocket associated with localhost and port 1234
+const socket = createQuicSocket({ endpoint: { port: 1234 } });
```

Comment:
Yes, port can be 0 using the same auto-select semantics we use elsewhere.

The relationship between `QuicEndpoint`, `QuicSocket`, and `QuicSession` is a bit complicated.

A `QuicSocket` is really a logical component that may be bound to one or more local UDP ports (`QuicEndpoint`). One of those endpoints is considered the default or `'preferred'` endpoint.

When a `QuicSocket` is told to act as a server (by calling `listen()`) it will bind to each of it's one or more configured endpoints and will listen on each for incoming initial QUIC packets.

When a `QuicSocket` is told to act as a client (by calling `connect()`) it will only use the preferred endpoint.

When a `QuicSession` is created, it has four pieces of critical information:

* It's Connection ID (the identifier this `QuicSession` instance has selected for the peer to reference it by... this is actually a set of one or more but let's forget that for now). This may change through the lifetime of the `QuicSession`.

* The peer's Connection ID (the identifier the QUIC peer selected for this `QuicSession` to reference it by). This may change through the lifetime of the `QuicSession`.

* The Path which a tuple consisting of the local udp IP and port and the remote udp IP and port. The path for a single `QuicSession` may change through the lifetime of the `QuicSession`

* A reference to the `QuicSocket` currently being used by the `QuicSession`. This also may change through the lifetime of the `QuicSession`.

When a `QuicEndpoint` receives a UDP packet, it forwards that packet on to it's owning `QuicSocket`, which checks the Connection ID to see if it identifies a `QuicSession` known to the `QuicSocket`. If a matching `QuicSession` is found, the packet is forwarded on to the `QuicSession` for handling. If a matching `QuicSession` is not found, we check to see if the `QuicSocket` is listening as a server and if the packet received is a valid initial packet. If both of those checks pass, a new `QuicSession` is created and the packet is forwarded.

When the `QuicSession` wants to send a QUIC packet, it forwards that packet to the `QuicSocket` along with the Path tuple. The `QuicSocket` determines which of it's set of `QuicEndpoint` instances matches the Path and forwards the packet on to the `QuicEndpoint` for transmission.

A single `QuicSocket`, then, can be bound to multiple local UDP ports and supports many client and server QUIC connections simultaneously. Also, a single `QuicSession` can receive or transmit over multiple UDP ports over it's lifetime and can be migrated to a completely different `QuicSocket` instance if necessary. It is even possible for the `QuicSession` to continue to exist in absence of any active network connectivity.